src/Workspaces/Core/Portable/Workspace/Host/PersistentStorage/IPersistentStorageLocationService.cs

```diff
@@ -53,11 +24,36 @@ public DefaultPersistentStorageLocationService()
         {
         }
 
-        public bool IsSupported(Workspace workspace) => false;
+        public virtual bool IsSupported(Workspace workspace) => false;
 
-        public string TryGetStorageLocation(SolutionId solutionId) => null;
-#pragma warning disable CS0067 // the event is unused
-        public event EventHandler<PersistentStorageLocationChangingEventArgs> StorageLocationChanging;
-#pragma warning disable CS0067
+        public string? TryGetStorageLocation(Solution solution)
+        {
+            if (!IsSupported(solution.Workspace))
+                return null;
+
+            if (string.IsNullOrWhiteSpace(solution.FilePath))
+                return null;
+
+            // Ensure that each unique workspace kind for any given solution has a unique
+            // folder to store their data in.
+
+            // Store in the LocalApplicationData/Roslyn/hash folder (%appdatalocal%/... on Windows,
+            // ~/.local/share/... on unix).  This will place the folder in a location we can trust
+            // to be able to get back to consistently as long as we're working with the same
+            // solution and the same workspace kind.
+            var appDataFolder = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData, Environment.SpecialFolderOption.Create);
+            var kind = StripInvalidPathChars(solution.Workspace.Kind ?? "");
+            var hash = StripInvalidPathChars(Checksum.Create(solution.FilePath).ToString());
+
+            return Path.Combine(appDataFolder, "Roslyn", "Cache", kind, hash);
```

Comment:
💡 Let's move this into the Visual Studio subfolder

```suggestion
            return Path.Combine(appDataFolder, "Microsoft", "VisualStudio", "Roslyn", "Cache", kind, hash);
```

Reply:
👍

Reply:
Love it.  Thanks!